### ISSUE TITLE
Generic server

### ISSUE THREAD

**Issue by phadej** [haskell-servant/servant#713]
I wrote in more detail in http://oleg.fi/gists/posts/2017-03-15-solution-to-servant-type-indentation-problem.html

Shorly:

```haskell
type API' =
   '[ "foo" :> Get '[JSON] Int
    , "bar" :> QueryParam "q" Text :> Get '[JSON] Text
    , "quu" :> ReqBody '[JSON] Text :> Post '[JSON] Bool
    ]

type API = ListApi API'

listApiServer
    :: forall a x xs.
       (Generic a, SListI xs, Code a ~ '[ServerMap (x ': xs)])
    => Proxy (x ': xs) -> a -> Server (ListApi (x ': xs))

data ApiImpl m = ApiImpl
    { apiFoo :: m Int
    , apiBar :: Maybe Text -> m Text
    , apiQuu :: Text -> m Bool
    }

deriveGeneric ''ApiImpl

check :: Code (ApiImpl Handler) :~: '[ServerMap API']
check = Refl

app :: Application
app = serve api $ listApiServer api' impl
  where
    impl :: ApiImpl Handler
    impl = ApiImpl {..}

    apiFoo = return 42

    apiBar Nothing  = return "empty"
    apiBar (Just x) = return x

    apiQuu t = return (t == "works?")
```

What you think, could we add these to `servant` (`ListApi`) and `servant-server` (`listApiServer`)?

**Comment by kosmikus** [comment]
Some comments after a very short first look:

- In general, I think having generic server functionality to match the generic client is very desirable.
- I think the generic server should not force you to write your API as a list. At the moment, it essentially does, because you need a proxy for the list-based version.
- Having utility functions that fold `:<|>` over a list of API types or flatten an API into a list of types (which we've been talking about before) is probably a good idea; but they should be as independent of the generic server functionality as possible.

**Comment by phadej** [comment]
I agree, that we should force the API as a list. I hope we can work around the proxy limitation, I didn't try too hard.

**Comment by alpmestan** [comment]
Does servant-flatten (and the type machinery it comes with) help in any way with this?

**Comment by domenkozar** [comment]
This can be closed as servant--server-0.14.1 has Servant.Server.Generic
